cli/command/cli.go

```diff
@@ -82,6 +82,7 @@ type DockerCli struct {
 	dockerEndpoint     docker.Endpoint
 	contextStoreConfig store.Config
 	initTimeout        time.Duration
+	userAgent          string
```

Comment:
- I recall I opened https://github.com/moby/moby/pull/45677 to add an option to the APIClient; wondering if we can punch through to that one, without _also_ having to store it in the DockerCLI

Reply:
`DockerCli` uses that helper for API client construction! But `DockerCli` also lazily creates `RegistryClient` and `NotaryClient` instances and passes along a user agent to them, so it does need it.

I agree that we might want to add a `WithAPIClientOption` passthrough for future tweaks to the API client without needing 2x PRs each time

Reply:
Ah! Right, yeah, I hate it that we _also_ re-invent our own "registry" client here (and I _think_ all that 💩  is for, erm, mostly legacy stuff).

I'll have a closer look if I can come up with some solutions (otherwise this looks fine to me). Things I still have my eyes on (not related to your changes per-se, but a bit of a sore point already);


- not a fan of requiring a fixed argument to pass around "user-agent" everywhere
- ☝️ even more if that is "hard-coded" (hard-coded, as in: a function-call that is hard-coded)
- ☝️ at least your PR improves _that_ part

But indeed persisting in 2 separate locations isn't ideal; also considering; is this something we should pass over the context (at least that would allow passing this around without requiring it to be passed as a dedicated argument that "somehow" needs to find its way somewhere)

Reply:
Clarification: when you say context, do you mean Docker context or Go `context.Context`?

It could canonically live inside the API client, but then _that_ interface would need to gain a getter so that `DockerCli` could read it back out to pass to registry & notary client.

Reply:
Doh! Sorry, somehow missed your last comment (also still need to look at this PR again 🙈)

I meant `context.Context`  and to (ab)use the context to pass this information, which would allow functions that need a user-agent to get access to it (which could potentially avoid having to punch-through a `UserAgent` argument to many places.

I haven't given that idea much thought yet though; I was considering that approach for API version (but for API version it may be more "clear-cut" that the API version to use would usually be "per request", and as such "per context"; for User-Agent that's more of a "grey area" I guess 😂